From b98da2135d8a0fdbcae37c842a318800c64abe4a Mon Sep 17 00:00:00 2001
From: Shivani Soni <quic_soni@quicinc.com>
Date: Wed, 27 Dec 2023 13:03:26 +0530
Subject: [PATCH] arm/arm64: enable hairpin for WLAN netdevices only.

enable hairpin for the WLAN netdevices.
For ethernet interfaces, hairpin not allowed
due to issues with switchdevs.

Change-Id: I72bcecfe52b2eac34bbb0d98249806bd8a82bf14
Signed-off-by: Shivani Soni <quic_soni@quicinc.com>
Signed-off-by: Karthik T S <quic_kartikts@quicinc.com>
---
 net/bridge/br_forward.c | 28 +++++++++++++++++++++++-----
 1 file changed, 23 insertions(+), 5 deletions(-)

diff --git a/net/bridge/br_forward.c b/net/bridge/br_forward.c
index 6fd7652aad87..68969e3edf83 100644
--- a/net/bridge/br_forward.c
+++ b/net/bridge/br_forward.c
@@ -24,11 +24,29 @@ static inline int should_deliver(const struct net_bridge_port *p,
 	struct net_bridge_vlan_group *vg;
 
 	vg = nbp_vlan_group_rcu(p);
-	return (((p->flags & BR_HAIRPIN_MODE) && !is_multicast_ether_addr(eth_hdr(skb)->h_dest))
-		|| skb->dev != p->dev) &&
-		(br_mst_is_enabled(p->br) || p->state == BR_STATE_FORWARDING) &&
-		br_allowed_egress(vg, skb) && nbp_switchdev_allowed_egress(p, skb) &&
-		!br_skb_isolated(p, skb);
+
+	if ((skb->dev != p->dev) &&
+	    br_allowed_egress(vg, skb) && p->state == BR_STATE_FORWARDING &&
+	    nbp_switchdev_allowed_egress(p, skb) &&
+	    !br_skb_isolated(p, skb)) {
+		return true;
+	}
+
+	if ((skb->dev == p->dev) &&
+	    (p->flags & BR_HAIRPIN_MODE) &&
+	    br_allowed_egress(vg, skb) && p->state == BR_STATE_FORWARDING) {
+
+		/* Allow hairpin only on WLAN netdevices.
+		 * For ethernet interfaces, hairpin not allowed
+		 * due to issue with switchdevs. */
+		if ( skb->dev->ieee80211_ptr == NULL) {
+			return false;
+		}
+
+		return true;
+	}
+
+	return false;
 }
 
 int br_dev_queue_push_xmit(struct net *net, struct sock *sk, struct sk_buff *skb)
-- 
2.34.1

